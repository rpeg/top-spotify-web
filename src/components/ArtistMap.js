import React from 'react';
import PropTypes from 'prop-types';
import {
  ComposableMap,
  Geographies,
  Geography,
  Sphere,
} from 'react-simple-maps';

import { makeSortedFrequencyArr } from '../lib/frequency';

const geoUrl = 'https://raw.githubusercontent.com/zcreativelabs/react-simple-maps/master/topojson-maps/world-110m.json';

const ArtistMap = ({ artistCountries }) => {
  const countries = artistCountries.map((a) => a.country);
  const sortedCountries = makeSortedFrequencyArr(countries);

  const max = sortedCountries
    .map((c) => c.freq)
    .reduce((acc, curr) => Math.max(acc, curr));

  return (
    <div
      className="svg-container"
      style={{
        display: 'inline-block',
        position: 'relative',
        width: '100%',
        paddingBottom: 'calc(100% * 4 / 8)',
        verticalAlign: 'top',
        overflow: 'hidden',
      }}
    >
      <ComposableMap
        className="svg-content"
        projectionConfig={{
          rotate: [-10, 0, 0],
          scale: 147,
        }}
        width={800}
        height={400}
        style={{ width: '100%', height: 'auto' }}
        viewBox="0 0 800 400"
      >
        <Sphere stroke="#1db954" strokeWidth={2} />
        <Geographies geography={geoUrl}>
          {({ geographies }) => geographies.map((geo) => {
            const match = sortedCountries.find(
              (c) => c.name.toLowerCase() === geo.properties.NAME_LONG.toLowerCase(),
            );

            return (
              <Geography
                key={geo.rsmKey}
                geography={geo}
                style={{
                  default: {
                    fill: `${match ? '#1db954' : '#282828'}`,
                    fillOpacity: `${match ? `${Math.max(match.freq / max, 0.25)}` : '0.5'}`,
                  },
                  hover: {
                    fill: '#D3D3D3',
                  },
                }}
              />
            );
          })}
        </Geographies>
      </ComposableMap>
    </div>


  );
};

ArtistMap.propTypes = {
  artistCountries: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default ArtistMap;
