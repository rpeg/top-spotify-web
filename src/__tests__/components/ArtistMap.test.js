import React from 'react';
import { shallow } from 'enzyme';

import ArtistMap from '../../components/ArtistMap';

const items = [
  { artist: 'Demilich', country: 'Finland' },
  { artist: 'Darkthrone', country: 'Norway' },
  { artist: 'Autopsy', country: 'United States' },
  { artist: 'Blasphemy', country: 'Canada' },
  { artist: 'Peste Noire', country: 'France' }];

describe('ArtistMap', () => {
  it('should not log errors in console', () => {
    const wrapper = shallow(<ArtistMap artistCountries={items} />);
    const spy = jest.spyOn(global.console, 'error');
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  });

  it('renders correctly', () => {
    const wrapper = shallow(<ArtistMap artistCountries={items} />);
    expect(wrapper).toMatchSnapshot();
  });
});
