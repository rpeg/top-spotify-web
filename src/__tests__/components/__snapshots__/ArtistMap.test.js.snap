// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtistMap renders correctly 1`] = `
<div
  className="svg-container"
  style={
    Object {
      "display": "inline-block",
      "overflow": "hidden",
      "paddingBottom": "calc(100% * 4 / 8)",
      "position": "relative",
      "verticalAlign": "top",
      "width": "100%",
    }
  }
>
  <ComposableMap
    className="svg-content"
    height={400}
    projectionConfig={
      Object {
        "rotate": Array [
          -10,
          0,
          0,
        ],
        "scale": 147,
      }
    }
    style={
      Object {
        "height": "auto",
        "width": "100%",
      }
    }
    viewBox="0 0 800 400"
    width={800}
  >
    <Memo(Sphere)
      stroke="#1db954"
      strokeWidth={2}
    />
    <Geographies
      geography="https://raw.githubusercontent.com/zcreativelabs/react-simple-maps/master/topojson-maps/world-110m.json"
    >
      <Component />
    </Geographies>
  </ComposableMap>
</div>
`;
